flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,211 @@
+        ReadWriteUtils.saveMetadata(this, path);
+        ReadWriteUtils.saveModelData(
+                KnnModelData.getModelDataStream(modelDataTable),
+                path,
+                new KnnModelData.ModelDataEncoder());
+    }
+
+    /**
+     * Loads model data from path.
+     *
+     * @param env Stream execution environment.
+     * @param path Model path.
+     * @return Knn model.
+     */
+    public static KnnModel load(StreamExecutionEnvironment env, String path) throws IOException {
+        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
+        KnnModel model = ReadWriteUtils.loadStageParam(path);
+        DataStream<KnnModelData> modelData =
+                ReadWriteUtils.loadModelData(env, path, new KnnModelData.ModelDataDecoder());
+        return model.setModelData(tEnv.fromDataStream(modelData));
+    }
+
+    /** This operator loads model data and predicts result. */
+    private static class PredictLabelFunction extends RichMapFunction<Row, Row> {
+        private final String featureCol;
+        private KnnModelData knnModelData;
+        private final int k;
+        private final String broadcastKey;
+
+        public PredictLabelFunction(String broadcastKey, int k, String featureCol) {
+            this.k = k;
+            this.broadcastKey = broadcastKey;
+            this.featureCol = featureCol;
+        }
+
+        @Override
+        public Row map(Row row) {
+            if (knnModelData == null) {
+                knnModelData =
+                        (KnnModelData)
+                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
+            }
+            DenseVector feature = (DenseVector) row.getField(featureCol);
+            double prediction = predict(feature);
+            return Row.join(row, Row.of(prediction));
+        }
+
+        private double predict(DenseVector inputFeature) {
+            PriorityQueue<Tuple2<Double, Double>> priorityQueue =
+                    new PriorityQueue<>(Comparator.comparingDouble(o -> -o.f0));
+            double featureNorm = 0.0;
+            for (int i = 0; i < inputFeature.size(); ++i) {
+                featureNorm += inputFeature.values[i] * inputFeature.values[i];
+            }
+            Tuple2<DenseVector, Double> sample = Tuple2.of(inputFeature, featureNorm);
+            double[] labelValues = knnModelData.labels.values;
+            double[] normValues = knnModelData.featureNorms.values;
+            DenseVector distanceVector = new DenseVector(labelValues.length);
+            BLAS.gemv(-2.0, knnModelData.packedFeatures, true, sample.f0, 0.0, distanceVector);
+            for (int i = 0; i < distanceVector.values.length; i++) {
+                distanceVector.values[i] =
+                        Math.sqrt(Math.abs(distanceVector.values[i] + sample.f1 + normValues[i]));
+            }
+            List<Tuple2<Double, Double>> distanceAndLabels = new ArrayList<>(labelValues.length);
```

Comment:
Is `distanceAndLabels` necessary here? We can directly access the `distanceVector` and `labels`, right?

Reply:
OK